### PR TITLE
fix(metrics): Light normalize before extracting metrics [INGEST-1517 INGEST-1424]

### PR DESCRIPTION
**Pull request by iker-barriocanal**
# Context

Currently, all relay modes extract metrics from transactions. However, only processing relays are performing normalization operations that sample out events and modify the values that relay extracts metrics from, in the store normalization step. Not sampling out events early enough also introduces the consequence of extracting metrics from transactions that are discarded a bit later.

# What this PR does

Starting from `envelopes.rs`, there's a new function `light_normalize_event` to perform minimal event normalization to filter them later, and that minimal normalization has been extracted from the processors.

The changes in the processors are mostly refactors extracting functions to make them callable from the outside (from `light_normalize_event`, in this case). The normalization processor doesn't re-do the normalization work that was previously done in the light processing step.

An alternative to all this extraction into functions is to have another processor. This was discarded because of its complexity. No additional design choices have been considered.

# Other considerations

Performing light normalization in non-processing relays impacts any test depending on the amount of time relay takes to process an envelope. The cause is the long amount of time relay takes to initialize lazily the regex to parse user agents. Since that work is only done once, it doesn't impact relay instances on production.

### CODE REVIEW

File: relay-general/src/store/normalize.rs [hunk 1]
```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
...
+            meta,
+            received_at,
+            max_secs_in_past,
+            max_secs_in_future,
+        )?;
+        normalize_event_tags(event)?;
+        normalize_exceptions(event)?;
+        normalize_user_agent(event, Some(true));
+        normalize_measurements(event);
+        normalize_breakdowns(event, breakdowns_config);
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: Would be nice to have a comment on each function call here, with the reason why it's needed for metrics extraction and/or inbound filters.

File: relay-general/src/store/normalize.rs [hunk 2]
```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
...
+
+            // In an ideal world all SDKs would set {{auto}} explicitly.
+            if let Some("javascript") | Some("cocoa") | Some("objc") = platform {
+                user.ip_address = Annotated::new(client_ip.to_owned());
+            }
         }
     }
 }
 
+pub fn light_normalize_event(
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs:512]: nit: We should probably come up with a better name for this function, but I don't have one yet. Maybe `pre_normalize_event`?

File: relay-server/src/actors/envelopes.rs [hunk 3]
```diff
@@ -1853,6 +1853,36 @@ impl EnvelopeProcessor {
...
+        light_normalize_event(
+            &mut state.event,
+            client_ip.as_ref(),
+            user_agent,
+            received_at,
+            max_secs_in_past,
+            max_secs_in_future,
+            breakdowns_config,
+        )
+        .map_err(|e| ProcessingError::ProcessingFailed(e))?;
```
Comment (jjbayer) [inline relay-server/src/actors/envelopes.rs]: We could wrap this call in a `metric!(timer(...))` to measure how long light normalization takes.

File: relay-server/src/actors/envelopes.rs [hunk 4]
```diff
@@ -1882,6 +1912,8 @@ impl EnvelopeProcessor {
 
             self.finalize_event(state)?;
 
+            self.light_normalize_event(state)?;
+
```
Comment (jjbayer) [inline relay-server/src/actors/envelopes.rs:1906]: With this, we can remove a lot of ad-hoc normalization from `extract_transaction_metrics`. But that can be a separate PR.

File: relay-general/src/store/normalize.rs [hunk 5]
```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
...
+    let site = std::mem::take(&mut event.site);
+    if site.value().is_some() {
+        tags.insert("site".to_string(), site);
+    }
+
+    Ok(())
+}
+
+/// Validates the timestamp range and sets a default value.
+fn normalize_timestamps(
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs:343]: A lot of these functions have been moved around, so it's hard to spot if any changes were made to them. Do you think we could leave them at their original position to reduce the diff?

If your intention was to make sure that functions are declared in the order that they are used, we can open a separate PR after this one that only reorders code.

File: relay-general/src/store/normalize.rs (hunk 1 above)
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: I quickly checked these functions for idempotency, and I think they all are, but ideally we would add unit tests to verify that calling these functions twice does not change the result.

File: relay-general/src/store/transactions.rs [hunk 6]
```diff
@@ -73,44 +116,7 @@ impl Processor for TransactionsProcessor {
...
-
-                trace_context.op.get_or_insert_with(|| "default".to_owned());
-            }
-            _ => {
-                return Err(ProcessingAction::InvalidTransaction(
-                    "context at event.contexts.trace must be of type trace.",
-                ));
-            }
-        }
+        validate_transaction(event)?;
```
Comment (jjbayer) [inline relay-general/src/store/transactions.rs:231]: I think you mentioned it in the PR description, but couldn't we remove this call, now that this is already called from `light_normalize_event`?

**Review by jjbayer — COMMENTED** [review]
This looks good overall! I did not yet verify whether everything that we need for inbound filters and / or metrics extraction is now actually part of `light_normalize_event`.

**Review by untitaker — COMMENTED** [review]
please check out sentry, do `pip install -e ~/projects/relay/py/` and run its tests. I think you may break some tests by moving normalization out of the store normalizer, which we use in Python to fully normalize events.

It may be easier to introduce a flag on StoreNormalizer to opt into "light normalization", but it all depends on how many tests you're going to break.

File: relay-general/src/store/normalize.rs (hunk 5 above)
Comment (jjbayer) [reply relay-general/src/store/normalize.rs:343]: Ignore my previous comment, I just realized these used to be methods and now they are standalone functions.

File: relay-general/src/store/normalize.rs (hunk 2 above)
Comment (iker-barriocanal) [reply relay-general/src/store/normalize.rs:512]: I hard agree "light normalization" isn't descriptive enough. I don't like `pre_normalize_*` because that indicates the operation happens _before_ normalization, and this function actually performs some normalization. I think that _some_ is what causes troubles for the naming. I'd also like to avoid referring to any filtering in the name, if possible -- the function may be used in another context in the future and shouldn't be tied up to current behavior.

What about something like `minimal_normalize_event`? That indicates there's some normalization happening, it doesn't perform the whole normalization, and the explanation of what it does and the motivation behind it could be added to the docstrings. What do you think?

File: relay-general/src/store/transactions.rs (hunk 6 above)
Comment (iker-barriocanal) [reply relay-general/src/store/transactions.rs:231]: That's a good question, and the motivation behind that decision isn't in the PR description. Normalizing the top-level transaction, in the current state of the PR, is indeed duplicated work. However, the processor also normalizes nested spans, and removing that validation step from the processor makes that function to not run on each span. I don't know and haven't investigated the consequences of doing that, but I don't want to take the risk there.

This explanation shouts for some refactoring to make the code easier to understand and not perform the duplicated work, but that is definitely out of scope for now.

**Comment by jjbayer** [comment]
> I think you may break some tests by moving normalization out of the store normalizer, which we use in Python to fully normalize events.

@untitaker This is a good point, should we just re-do everything in store normalization, and for the time being accept the additional cost for the sake of correctness?

File: relay-general/src/store/transactions.rs (hunk 6 above)
Comment (jjbayer) [reply relay-general/src/store/transactions.rs:231]: But the normalization of the spans happens on lines 121 - 131, right? The function `validate_transaction` does not touch spans as far as I can see, but I might be missing something.

Anyway I think we can leave this as-is, just to be extra sure, but I would add a code comment here that this is a candidate for refactoring.

File: relay-general/src/store/normalize.rs [hunk 7]
```diff
@@ -291,166 +158,352 @@ impl<'a> NormalizeProcessor<'a> {
...
-
-            if !headers.contains("User-Agent") {
-                headers.insert(
-                    HeaderName::new("User-Agent"),
-                    Annotated::new(HeaderValue::new(client.clone())),
-                );
-            }
-        }
+        #[cfg(not(feature = "uaparser"))]
+        panic!("relay not built with uaparser feature");
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: This looks risky, shall we just remove the `optional` in Cargo.toml and get rid of the feature gates?

https://github.com/getsentry/relay/blob/d91984ff4327973cb4f97b943736e834e1ee2b4d/relay-general/Cargo.toml#L38

File: relay-cabi/src/processing.rs [hunk 8]
```diff
@@ -151,6 +155,8 @@ pub unsafe extern "C" fn relay_pii_strip_event(
     let mut processor = PiiProcessor::new(&compiled);
 
     let mut event = Annotated::<Event>::from_json((*event).as_str())?;
+    validate_annotated_transaction(&mut event)?;
+    light_normalize_event(&mut event, None, None, None, None, None, None)?;
```
Comment (jjbayer) [inline relay-cabi/src/processing.rs]: Why add light normalization here? `processor` at this point is a `PiiProcessor`, not a `StoreProcessor`, so the call to `process_value` is not related to store normalization.

File: tests/integration/test_security_report.py [hunk 9]
```diff
@@ -126,14 +126,19 @@ def test_security_report(mini_sentry, relay, test_case, json_fixture_provider):
...
         event.pop(x, None)
 
     ext = ".no_processing.output"
     expected_evt = fixture_provider.load(test_name, ext)
 
+    if "received" in event:
+        event.pop("received")
+    if "timestamp" in event:
+        event.pop("timestamp")
+
```
Comment (iker-barriocanal) [inline tests/integration/test_security_report.py:141]: At least in the `expect_ct` case the event has the `received` and `timestamp` fields. I haven't thought if we want them, but I don't think this is that important right now and I took the short and hacky path deleting these fields here. We may want to revisit this.

**Review by iker-barriocanal — COMMENTED** [review]
There's a decent amount of stuff added to the output of integration tests of security reports. By looking at the code, all changes seem they make sense to me, so I assume they weren't running before.

File: tests/integration/test_security_report.py [hunk 10]
```diff
@@ -46,7 +46,7 @@ def test_uses_origins(mini_sentry, relay, json_fixture_provider, allowed_origins
     )
 
     if should_be_allowed:
-        mini_sentry.captured_events.get(timeout=1).get_event()
+        mini_sentry.captured_events.get(timeout=10).get_event()
```
Comment (jjbayer) [inline tests/integration/test_security_report.py:49]: Note: This is necessary because the non-processing relay now does normalization, which includes UA parsing, which apparently has slow initialization.

File: CHANGELOG.md [hunk 11]
```diff
@@ -15,7 +15,7 @@
 - Generate mobile measurements frames_frozen_rate, frames_slow_rate, stall_percentage. ([#1373](https://github.com/getsentry/relay/pull/1373))
 - Change to the internals of the healthcheck endpoint. ([#1374](https://github.com/getsentry/relay/pull/1374))
 - Re-encode the Typescript payload to normalize. ([#1372](https://github.com/getsentry/relay/pull/1372))
-
+- Light normalize before extracting metrics. ([#1366](https://github.com/getsentry/relay/pull/1366))
```
Comment (jjbayer) [inline CHANGELOG.md]:
```suggestion
- Partially normalize events before extracting metrics. ([#1366](https://github.com/getsentry/relay/pull/1366))
```

File: relay-general/src/store/normalize.rs [hunk 12]
```diff
@@ -907,6 +939,7 @@ fn test_user_ip_from_remote_addr() {
 
     let config = StoreConfig::default();
     let mut processor = NormalizeProcessor::new(Arc::new(config), None);
+    light_normalize(&mut event, None, None, None, None, None, None).unwrap();
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: nit: We should really pass a config struct instead of passing this many arguments, but let's leave that refactor for a follow-up PR.

File: relay-general/src/store/mod.rs [hunk 13]
```diff
@@ -137,3 +140,25 @@ impl<'a> Processor for StoreProcessor<'a> {
...
+    event: &mut Annotated<Event>,
+    client_ip: Option<&IpAddr>,
+    user_agent: Option<&'a str>,
+    received_at: Option<DateTime<Utc>>,
+    max_secs_in_past: Option<i64>,
+    max_secs_in_future: Option<i64>,
+    breakdowns_config: Option<BreakdownsConfig>,
+) -> ProcessingResult {
+    validate_annotated_transaction(event)?;
+    light_normalize_event(
```
Comment (jjbayer) [inline relay-general/src/store/mod.rs]: nit: If my search is correct, `light_normalize_event` is never called outside of `light_normalize`. So why not simplify this by calling `validate_annotated_transaction` from inside `light_normalize_event` (and then remove the `light_normalize` wrapper)?

**Review by jjbayer — APPROVED** [review]

File: tests/integration/test_security_report.py (hunk 9 above)
Comment (jjbayer) [reply tests/integration/test_security_report.py:141]: I think this is fine, just a result of timestamp normalization.

File: relay-general/src/store/mod.rs (hunk 13 above)
Comment (iker-barriocanal) [reply relay-general/src/store/mod.rs]: Good feedback, I'll do that in a follow-up PR.

File: relay-general/src/store/mod.rs [hunk 14]
```diff
@@ -8,7 +8,9 @@ use serde_json::Value;
 
 use crate::processor::{ProcessingState, Processor};
 use crate::protocol::{Event, IpAddr};
-use crate::types::{Meta, ProcessingResult, SpanAttribute};
+use crate::types::{Annotated, Meta, ProcessingResult, SpanAttribute};
+use normalize::light_normalize_event;
+use transactions::validate_annotated_transaction;
```
Comment (jan-auer) [inline relay-general/src/store/mod.rs]: (before merging) Please qualify functions at use site.

File: relay-general/src/store/mod.rs [hunk 15]
```diff
@@ -137,3 +140,25 @@ impl<'a> Processor for StoreProcessor<'a> {
...
 }
+
+pub fn light_normalize<'a>(
+    event: &mut Annotated<Event>,
+    client_ip: Option<&IpAddr>,
+    user_agent: Option<&'a str>,
+    received_at: Option<DateTime<Utc>>,
+    max_secs_in_past: Option<i64>,
+    max_secs_in_future: Option<i64>,
+    breakdowns_config: Option<BreakdownsConfig>,
```
Comment (jan-auer) [inline relay-general/src/store/mod.rs]: (before merging) These are now enough parameters to warrant some form of config struct like we have in the store normalizer.

File: relay-general/src/store/mod.rs (hunk 13 above)
Comment (jan-auer) [reply relay-general/src/store/mod.rs]: +1 on this, would even prefer if this goes into the same PR.

File: relay-server/src/actors/envelopes.rs (hunk 4 above)
Comment (jan-auer) [reply relay-server/src/actors/envelopes.rs:1906]: Are we sure that double-normalizing (such as computing the breakdowns) doesn't break anything now?

File: relay-general/src/store/normalize.rs [hunk 16]
```diff
@@ -344,166 +211,355 @@ impl<'a> NormalizeProcessor<'a> {
...
+            if protocol::validate_environment(environment).is_ok() {
+                Ok(())
+            } else {
+                meta.add_error(ErrorKind::InvalidData);
+                Err(ProcessingAction::DeleteValueSoft)
+            }
+        })?;
+
+        // Default required attributes, even if they have errors
+        normalize_release_dist(event); // dist is a tag extracted along with other metrics from transactions
```
Comment (jan-auer) [inline relay-general/src/store/normalize.rs:545]: _nit_: Do we need all these comments here?

**Review by jan-auer — COMMENTED** [review]
I believe, we need to run the `SchemaProcessor` at the very least since that enforces things like: `max_chars`, `required`, `trim_whitespace`, and `nonempty`. In store normalization, the schema processor runs before the normalizer, which means that it can change values in such a way that the normalizer behaves differently.

File: relay-general/src/store/normalize.rs (hunk 16 above)
Comment (jjbayer) [reply relay-general/src/store/normalize.rs:545]: I asked for these comments because I figured it might be helpful to know in the future why we decided to pull those into "light" normalization.

File: relay-general/src/store/normalize.rs [hunk 17]
```diff
@@ -823,6 +830,18 @@ impl Default for NormalizeProcessor<'_> {
...
+fn get_empty_light_normalization_config<'a>() -> LightNormalizationConfig<'a> {
+    LightNormalizationConfig {
+        client_ip: None,
+        user_agent: None,
+        received_at: None,
+        max_secs_in_past: None,
+        max_secs_in_future: None,
+        breakdowns_config: None,
+    }
+}
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs:843]: nit: You can put `#[derive(default)]` above `struct LightNormalizationConfig`, and then call `Default::default()` to get an empty config.

**Review by jjbayer — APPROVED** [review]
Feel free to ignore nitpick.

**Review by untitaker — APPROVED** [review]

File: relay-general/src/store/mod.rs (hunk 13 above)
Comment (iker-barriocanal) [reply relay-general/src/store/mod.rs]: Addressed in https://github.com/getsentry/relay/pull/1385.

File: relay-general/src/store/normalize.rs (hunk 17 above)
Comment (iker-barriocanal) [reply relay-general/src/store/normalize.rs:843]: Addressed in https://github.com/getsentry/relay/pull/1386.
